web/src/unread_ops.ts

```diff
@@ -254,33 +263,77 @@ function process_newly_read_message(
-    let message_ids_to_update;
-    if (message_lists.current.data.fetch_status.has_found_newest()) {
-        message_ids_to_update = message_lists.current
-            .all_messages()
-            .filter((msg) => msg.id >= message_id)
-            .map((msg) => msg.id);
+    const has_found_newest = message_lists.current.data.fetch_status.has_found_newest();
+    const is_in_interleaved_view = current_filter.is_interleaved_view();
+    let message_ids_to_update: number[] | undefined;
+
+    function do_mark_unread(): void {
+        // If we have already fully fetched the current view, we can
+        // send the server the set of IDs to update, rather than
+        // updating on the basis of the narrow.
+        if (message_ids_to_update !== undefined && message_ids_to_update.length < 200) {
+            do_mark_unread_by_ids(message_ids_to_update);
+        } else {
+            const include_anchor = true;
+            const messages_marked_unread_till_now = 0;
+            const num_after = INITIAL_BATCH_SIZE - 1;
+            do_mark_unread_by_narrow(
+                message_id,
+                include_anchor,
+                messages_marked_unread_till_now,
+                num_after,
+                narrow,
+            );
+        }
     }
 
-    if (message_ids_to_update !== undefined && message_ids_to_update.length < 200) {
-        do_mark_unread_by_ids(message_ids_to_update);
+    if (!has_found_newest && !is_in_interleaved_view) {
+        do_mark_unread();
+        return;
+    }
+
+    const min_message_ids_to_update = message_lists.current
+        .all_messages()
+        .filter((msg) => msg.id >= message_id && !msg.unread)
+        .map((msg) => msg.id);
+    message_ids_to_update = has_found_newest ? min_message_ids_to_update : undefined;
+
+    if (
+        !is_in_interleaved_view ||
+        (message_ids_to_update !== undefined &&
+            message_ids_to_update.length < MIN_MARK_AS_UNREAD_COUNT_KNOWN)
+    ) {
+        do_mark_unread();
     } else {
-        const include_anchor = true;
-        const messages_marked_unread_till_now = 0;
-        const num_after = INITIAL_BATCH_SIZE - 1;
-        do_mark_unread_by_narrow(
-            message_id,
-            include_anchor,
-            messages_marked_unread_till_now,
-            num_after,
-            narrow,
-        );
+        const message_count = min_message_ids_to_update.length;
+        let display_count: string;
+
+        if (has_found_newest) {
+            display_count = message_count.toString();
+        } else {
+            const rounded_count =
+                Math.floor(message_count / UNREAD_COUNT_STEP_SIZE) * UNREAD_COUNT_STEP_SIZE;
+            const count_to_use =
+                message_count < UNREAD_COUNT_STEP_SIZE ? message_count : rounded_count;
+            display_count = `${count_to_use}+`;
+        }
+
+        const ctx = {
+            is_count_too_low: message_count < MIN_MARK_AS_UNREAD_COUNT_LOWER_BOUND,
```

Comment:
I think `show_message_count` would be a better name for this. `count` and `ctx` would also be better names than `N` and `ctx`; we don't use abbreviations in the codebase.